### PR TITLE
[DNMY] Opsd renewable ppls

### PR DESCRIPTION
**Pull request by eb5194**
Closes #192 

## Changes proposed in this Pull Request

add renewable assets given in a OPSD data publication to the network. Data is downloaded in powerplantmatching, so if new countries become available, they have to be added there.

If estimate_renewable_capacities is applied, then available renewable capacities are overwritten for those countries where precise lat and lon are given (which at the moment is false for e.g. Poland).

**works only with the latest version of powerplantmatching**. (19.10.2020 ?)

**updated PR of #192 because target branch was not `master`**

## Checklist

- [x] I tested my contribution locally and it seems to work fine.
- [x] Code and workflow changes are sufficiently documented.
- [x] Newly introduced dependencies are added to `environment.yaml` and `environment.docs.yaml`.
- [x] Changes in configuration options are added in all of `config.default.yaml`, `config.tutorial.yaml`, and `test/config.test1.yaml`.
- [x] Changes in configuration options are also documented in `doc/configtables/*.csv` and line references are adjusted in `doc/configuration.rst` and `doc/tutorial.rst`.
- [x] A note for the release notes `doc/release_notes.rst` is amended in the format of previous release notes.

### CODE REVIEW

**Comment by fneum** [comment]
ready for review @eb5194?

**Comment by fneum** [comment]
> works only with the latest version of powerplantmatching. (19.10.2020 ?)

DNMY is because https://github.com/FRESNA/powerplantmatching/commit/38b075def3d866be226b97fca7cc4805242ec108 is not released yet.

Needs a future release 0.4.6 of `powerplantmatchin` which should be reflected in the `environment.*.yaml` files @eb5194.

**Comment by eb5194** [comment]
> ready for review @eb5194?

for me it's working as it is, so yes. thanks!

**Comment by FabianHofmann** [comment]
Just let me update the ppm code. I will introduce some functionality to make this more slim

File: scripts/add_electricity.py [hunk 1]
```diff
@@ -476,7 +481,72 @@ def attach_extendable_generators(n, costs, ppl):
                                       "'{tech}' is not implemented, yet. "
                                       "Only OCGT, CCGT and nuclear are allowed at the moment.")
 
+def map_ppls_to_bus(xy, buses, tech):
+    # map ppl with coordinates xy \in R^2 to closest bus (in buses) of the network;
+    idx = np.sqrt((n.buses.loc[buses].x-xy[1])**2+(n.buses.loc[buses].y-xy[0])**2).argmin()
+    return n.buses.loc[buses].iloc[idx].name + ' ' + tech
```
Comment (FabianHofmann) [inline scripts/add_electricity.py:487]: This is better done using the `scipy.spatial.KDTree`

File: scripts/add_electricity.py [hunk 2]
```diff
@@ -476,7 +481,72 @@ def attach_extendable_generators(n, costs, ppl):
...
+        countries = (snakemake.config['electricity']
+                     .get('OPSD_VRES_countries', {}))
+
+    if len(countries) == 0: return
+
+    for country in countries:
+        mapping = attach_OPSD_ppls_for_country(country)
+        n.generators.loc[mapping.index, 'p_nom'] = mapping
+        logger.info('overwriting renewable capacities in {} [MW] with precise locations\n{}'
+                    .format(country, n.generators.loc[mapping.index].groupby('carrier').sum().p_nom))
```
Comment (FabianHofmann) [inline scripts/add_electricity.py:547]:
```suggestion
def attach_OPSD_renewables(n):
    available = ['DE', 'FR', 'PL', 'CH', 'DK', 'CZ', 'SE', 'GB']
    countries = set(available) & set(n.buses.country)
    tech_map = (snakemake.config['electricity']
                 .get('include_renewable_capacities_from_OPSD', {}))
    logger.info(f'Using OPSD renewable capacities in {", ".join(countries)}.')

    df = pd.concat([pm.data.OPSD_VRE_country(c) for c in countries])
    df = df.query('Fueltype in @techs').powerplant.convert_country_to_alpha2()
    technology_b = ~df.Technology.isin(['Onshore', 'Offshore'])
    df['Fueltype'] = df.Fueltype.where(technology_b, df.Technology)

    for fueltype, carrier_like in tech_map.items():
        gens = n.generators[lambda df: df.carrier.str.contains(carrier_like)]
        buses = n.buses.loc[gens.bus.unique()]
        gens_per_bus = gens.groupby('bus').p_nom.count()

        caps = map_country_bus(df.query('Fueltype == @fueltype'), buses)
        caps = caps.groupby(['bus']).Capacity.sum()

        n.generators.p_nom.update(gens.bus.map(caps).dropna())
```

Could you try out this function? (uses the current ppm master). It should include all the functionality from above

File: scripts/add_electricity.py [hunk 3]
```diff
@@ -476,7 +481,72 @@ def attach_extendable_generators(n, costs, ppl):
...
+            for tech in techs:
+                tech_i = n.generators.query('carrier in @tech and bus in @buses').index
+
+                # fix available buses in map_ppls_to_bus function:
+                map_ppls = partial(map_ppls_to_bus, buses=n.generators.loc[tech_i].bus, tech=tech)
+
+                # parallel assignment of ppls to buses; each process is independent to speed up
+                pool_buses = multiprocessing.Pool(processes=4)
+                mapping = (mapping.append(
+                    pd.Series((capacities.query('Fueltype in @ppm_fueltype').Capacity/len(techs)).values,
```
Comment (FabianHofmann) [inline scripts/add_electricity.py:523]: Here you divide by `len(techs)`. This assumes that all techs are at all buses, right?. However, there are cases where you have only onwind or only offwind (not both) at a bus.

File: scripts/add_electricity.py [hunk 4]
```diff
@@ -476,7 +481,72 @@ def attach_extendable_generators(n, costs, ppl):
...
+        mapping = pd.Series(dtype='float')
+        for ppm_fueltype, techs in tech_map.items():
+            for tech in techs:
+                tech_i = n.generators.query('carrier in @tech and bus in @buses').index
+
+                # fix available buses in map_ppls_to_bus function:
+                map_ppls = partial(map_ppls_to_bus, buses=n.generators.loc[tech_i].bus, tech=tech)
+
+                # parallel assignment of ppls to buses; each process is independent to speed up
+                pool_buses = multiprocessing.Pool(processes=4)
```
Comment (FabianHofmann) [inline scripts/add_electricity.py:521]: The pooling is nicely implemented, but when using `KDTree` (which is super fast) it is not necessary.

File: scripts/add_electricity.py [hunk 5]
```diff
@@ -476,7 +481,72 @@ def attach_extendable_generators(n, costs, ppl):
...
+                              (pool_buses.map(map_ppls,
+                                              np.array(capacities.query('Fueltype in @ppm_fueltype')[['lat','lon']]))))
+                    .groupby(level=0).sum())) # aggregate capacities per bus with groupby
+                
+        logger.info('Adding {} renewable assets in {} from OPSD'
+                .format(len(capacities), country))
+        return mapping
+    else: # if locations are not given, fall back to heuristics
+        logger.warning('lat and lon are not given for assets in {}; falling back to heuristics'
+                .format(country))
```
Comment (FabianHofmann) [inline scripts/add_electricity.py:533]: not necessarily right?

**Review by FabianHofmann — CHANGES_REQUESTED** [review]

**Comment by FabianHofmann** [comment]
I'm closing this in the favour of #212

**Pull request closed without merging**